src/Backend/SgateShopgatePlugin/Controllers/Frontend/shopgate.php

```diff
@@ -934,60 +1068,201 @@ public function loginUserAction()
-
-        $this->Response()->setBody(json_encode(array(
-            'id' => $user['customernumber'],
-            'mail' => $user['email'],
-            'firstName' => $user['firstname'],
-            'lastName' => $user['lastname'],
-            'birthday' => $user['birthday'],
-            'customerGroups' => $user['customergroup'],
-            'addresses' => array()
-        )));
-        $this->Response()->sendResponse();
-        exit();
+            $this->Response()->sendResponse();
+            exit();
+        }
     }
 
     /**
-     * Custom registration action which checks whether the user is inside the app's web view
+     * Custom action to update user data
      */
-    public function registrationAction()
+    public function updateUserAction()
     {
-        $path = $this->Request()->getPathInfo();
-        $segments = explode('/', trim($path, '/'));
+        $this->Response()->setHeader('Content-Type', 'application/json');
 
-        $sessionId = $segments[2];
+        $response = array(
+            'success' => true,
+            'message' => ''
+        );
 
-        if (isset($sessionId)) {
-            session_write_close();
-            session_id($sessionId);
+        try {
+            $params = $this->getJsonParams();
+            $decoded = $this->getJWT($params['token']);
+
+            $customer = $this->getCustomer($decoded['customer_id']);
+            $customer->setFirstname($decoded['first_name']);
+            $customer->setLastname($decoded['last_name']);
+            $customer->setAttribute($decoded['custom_attributes']);
+
+            Shopware()->Models()->persist($customer);
+            Shopware()->Models()->flush();
+
+            $response['success'] = true;
+            $response['message'] = $decoded['email'];
+
+        } catch (Exception $error) {
+            $response['message'] = $error->getMessage();
+        }
+
+        $this->Response()->setBody(json_encode($response));
+        $this->Response()->sendResponse();
+        exit();
+    }
+
+    /**
+     * Custom action to update user email
+     */
+    public function updateUserEmailAction()
+    {
+        if (!$this->Request()->isPut()) {
+            return;
+        }
+
+        $this->Response()->setHeader('Content-Type', 'application/json');
+
+        $response = array(
+            'success' => false,
+            'message' => ''
+        );
+
+        try {
+            $params = $this->getJsonParams();
+            $decoded = $this->getJWT($params['token']);
+
+            $customer = $this->getCustomer($decoded['customer_id']);
+
+            $form = $this->createForm("Shopware\\Bundle\\AccountBundle\\Form\Account\\EmailUpdateFormType", $customer);
+            $emailData = array(
+                'email' => $decoded['email'],
+                'emailConfirmation' => $decoded['email']
+            );
+            $form->submit($emailData, false);
+
+            if ($form->isValid()) {
```

Comment:
again duplicate code, just make sure to re-use code instead of copy/pasting. See function updateUserPasswordAction, lines 1234-1245